src/d3d11/d3d11_swapchain.cpp

```diff
@@ -241,59 +241,89 @@ namespace dxvk {
     // Flush pending rendering commands before
     auto immediateContext = static_cast<D3D11ImmediateContext*>(deviceContext.ptr());
     immediateContext->Flush();
-
-    // Wait for the sync event so that we respect the maximum frame latency
-    uint64_t frameId = ++m_frameId;
-    m_frameLatencySignal->wait(frameId - GetActualFrameLatency());
     
+    // Present this image, then acquire our next image for
+    // a future present.
+    // This means that Present will always present something
+    // instantly when called. Rather than waiting around
+    // to acquire an image.
+
     for (uint32_t i = 0; i < SyncInterval || i < 1; i++) {
       SynchronizePresent();
 
       if (!m_presenter->hasSwapChain())
         return DXGI_STATUS_OCCLUDED;
 
+      // If we don't have an acquired image and we can't
+      // acquire one, that means we've been occluded.
+      if (m_nextImageIndex == UINT32_MAX && !AcquireNextImage())
+        return DXGI_STATUS_OCCLUDED;
+
       // Presentation semaphores and WSI swap chain image
       vk::PresenterInfo info = m_presenter->info();
-      vk::PresenterSync sync = m_presenter->getSyncSemaphores();
-
-      uint32_t imageIndex = 0;
-
-      VkResult status = m_presenter->acquireNextImage(
-        sync.acquire, VK_NULL_HANDLE, imageIndex);
-
-      while (status != VK_SUCCESS && status != VK_SUBOPTIMAL_KHR) {
-        RecreateSwapChain(m_vsync);
-
-        if (!m_presenter->hasSwapChain())
-          return DXGI_STATUS_OCCLUDED;
-        
-        info = m_presenter->info();
-        sync = m_presenter->getSyncSemaphores();
-
-        status = m_presenter->acquireNextImage(
-          sync.acquire, VK_NULL_HANDLE, imageIndex);
-      }
 
       // Resolve back buffer if it is multisampled. We
       // only have to do it only for the first frame.
       m_context->beginRecording(
         m_device->createCommandList());
       
       m_blitter->presentImage(m_context.ptr(),
-        m_imageViews.at(imageIndex), VkRect2D(),
+        m_imageViews.at(m_nextImageIndex), VkRect2D(),
         m_swapImageView, VkRect2D());
 
       if (m_hud != nullptr)
         m_hud->render(m_context, info.format, info.imageExtent);
       
       if (i + 1 >= SyncInterval)
-        m_context->signal(m_frameLatencySignal, frameId);
+        m_context->signal(m_frameLatencySignal, m_frameId);
 
-      SubmitPresent(immediateContext, sync, i);
+      SubmitPresent(immediateContext, m_nextPresenterSync, i);
     }
 
+    // Signal the frame latency event for *this* frame.
     SignalFrameLatencyEvent();
-    return S_OK;
+
+    // Wait on the frame latency for the subsequent frame
+    // to avoid any latency at the next presentation.
+    // Bump our frame id.
+    WaitForFrameLatency(++m_frameId);
```

Comment:
This will immediately synchronize if frame latency is 1, rather than allow for a single frame to be queued.